### PR TITLE
refactor: implement repeat as a normalization to reduce number of models

### LINKED ISSUE #4947: Refactor Repeat as a Macro?
Our repeat should have the same behavior as using fold (unpivot) and then facet into row/column. We should investigate whether it makes sense to clean up the implementation and implement repeat as a macro.

### PR DESCRIPTION
**Pull request by domoritz**
Fixes #4947

This also fixes the layer type, which incorrectly extended generic layer spec.

### CODE REVIEW

**Comment by domoritz** [comment]
Rebased to master. Looking pretty good now.

File: examples/compiled/geo_repeat.vg.json [hunk 1]
```diff
@@ -85,128 +83,152 @@
   ],
   "projections": [
     {
-      "name": "projection",
-      "size": {"signal": "[width, height]"},
-      "fit": {
-        "signal": "[child__repeat_row_population_geojson_0, child__repeat_row_engineers_geojson_0, child__repeat_row_hurricanes_geojson_0]"
-      },
+      "name": "__row_population_projection",
```
Comment (kanitw) [inline examples/compiled/geo_repeat.vg.json]: projection doesn't get merged correctly.

File: examples/compiled/geo_repeat.vg.json [hunk 2]
```diff
@@ -30,7 +28,7 @@
         {
           "type": "geojson",
           "geojson": "geo",
-          "signal": "child__repeat_row_population_geojson_0"
+          "signal": "__row_population_geojson_0"
```
Comment (kanitw) [inline examples/compiled/geo_repeat.vg.json]: To minimize review churns, I'd rather have consistent naming pattern with the past. You can send another PR to change / simplify it later, but avoid it in this giant PR.

File: examples/compiled/geo_repeat.vg.json [hunk 3]
```diff
@@ -85,128 +84,158 @@
   ],
   "projections": [
     {
-      "name": "projection",
-      "size": {"signal": "[width, height]"},
-      "fit": {
-        "signal": "[child__repeat_row_population_geojson_0, child__repeat_row_engineers_geojson_0, child__repeat_row_hurricanes_geojson_0]"
-      },
+      "name": "__repeat_row_population_projection",
```
Comment (domoritz) [inline examples/compiled/geo_repeat.vg.json]: Projections are not merged because layout sizes are not merged correctly here.

Comment (kanitw) [reply examples/compiled/geo_repeat.vg.json]: Then fix the layout size to merge correctly? 

I don't think we should merge this without fixing it.

File: examples/compiled/geo_repeat.vg.json [hunk 4]
```diff
@@ -30,7 +29,7 @@
         {
           "type": "geojson",
           "geojson": "geo",
-          "signal": "child__repeat_row_population_geojson_0"
+          "signal": "__repeat_row_population_geojson_0"
```
Comment (kanitw) [inline examples/compiled/geo_repeat.vg.json]: The name is still changing.

File: examples/compiled/geo_repeat.vg.json (hunk 3 above)
Comment (domoritz) [reply examples/compiled/geo_repeat.vg.json]: Filed as https://github.com/vega/vega-lite/issues/6024 since it's orthogonal

File: examples/compiled/geo_repeat.vg.json [hunk 5]
```diff
@@ -85,22 +84,40 @@
   ],
   "projections": [
     {
-      "name": "projection",
-      "size": {"signal": "[width, height]"},
-      "fit": {
-        "signal": "[child__repeat_row_population_geojson_0, child__repeat_row_engineers_geojson_0, child__repeat_row_hurricanes_geojson_0]"
-      },
+      "name": "child__repeat_row_population_projection",
```
Comment (domoritz) [inline examples/compiled/geo_repeat.vg.json]: These changes come from an orthogonal issue: https://github.com/vega/vega-lite/issues/6024

File: examples/compiled/geo_repeat.vg.json (hunk 4 above)
Comment (domoritz) [reply examples/compiled/geo_repeat.vg.json]: Fixed

**Comment by kanitw** [comment]
Please rebase after https://github.com/vega/vega-lite/pull/6026 is merged.

**Comment by domoritz** [comment]
TODO:
* [ ] Make sure we still use the repeat config
* [ ] Clean up `extractCompositionLayout`

**Comment by domoritz** [comment]
Merge after https://github.com/vega/vega-lite/pull/6029

**Comment by domoritz** [comment]
Okay, I decided to just drop support for repeat configs. I don't think anyone really used it anyway. We could hack it through style maybe but I think this is an acceptable exception.

File: examples/compiled/interactive_panzoom_splom.vg.json [hunk 6]
```diff
@@ -118,22 +118,25 @@
   "marks": [
     {
       "type": "group",
-      "name": "child__repeat_row_Horsepower__repeat_column_Miles_per_Gallon_group",
+      "name": "child__repeat_row_Horsepowerchild__repeat_column_Miles_per_Gallon_group",
```
Comment (kanitw) [inline examples/compiled/interactive_panzoom_splom.vg.json]: The name is a bit weird here

**Comment by kanitw** [comment]
I think you just caused even more renames lol

**Comment by domoritz** [comment]
I think it's only these ones, right? 

Screen Shot 2020-03-08 at 18 28 00

**Comment by kanitw** [comment]
> I think it's only these ones, right?

IDK, because I haven't looked at them -- too many.

**Comment by domoritz** [comment]
I looked through the generated SVGs and this looks good to me. Anything left @kanitw?

File: examples/compiled/interactive_layered_crossfilter_discrete.vg.json [hunk 7]
```diff
@@ -183,6 +182,7 @@
   ],
   "signals": [
     {"name": "childWidth", "value": 200},
+    {"name": "childHeight", "value": 200},
```
Comment (kanitw) [inline examples/compiled/interactive_layered_crossfilter_discrete.vg.json]: Why does this become childHeight?

**Comment by kanitw** [comment]
> I looked through the generated SVGs and this looks good to me. Anything left @kanitw?

It'd be great to look at generated Vega too. 

Just added one comment after looking at a few Vega specs.

File: test/compile/repeat.test.ts [hunk 8]
```diff
@@ -1,10 +1,8 @@
-import {replaceRepeaterInEncoding} from '../../src/compile/repeater';
+import {replaceRepeaterInEncoding, replaceRepeaterInFacet} from '../../src/compile/repeater';
```
Comment (kanitw) [inline test/compile/repeat.test.ts]: Why does this file remains here at all given there is no model left?

File: src/spec/map.ts [hunk 9]
```diff
@@ -76,18 +88,19 @@ export abstract class SpecMapper<
...
       spec: this.map(spec.spec, params) as any
     };
   }
 
-  protected mapRepeat(spec: GenericRepeatSpec<UI, LI>, params: P): GenericRepeatSpec<UO, GenericLayerSpec<UO>> {
+  protected mapRepeat(spec: RepeatSpec, params: P): GenericSpec<UO, any, RO, FO> {
     return {
       ...spec,
-      spec: this.map(spec.spec, params)
+      spec: this.map(spec.spec as any, params)
```
Comment (kanitw) [inline src/spec/map.ts]: comment why `as any` here (even if it's the same as in `mapFacet` -- they are two different functions).

File: src/log/message.ts [hunk 10]
```diff
@@ -85,13 +85,9 @@ export function columnsNotSupportByRowCol(type: 'facet' | 'repeat') {
...
 
-// CONCAT
+// CONCAT / REPEAT
 export const CONCAT_CANNOT_SHARE_AXIS =
-  'Axes cannot be shared in concatenated views yet (https://github.com/vega/vega-lite/issues/2415).';
-
-// REPEAT
-export const REPEAT_CANNOT_SHARE_AXIS =
-  'Axes cannot be shared in repeated views yet (https://github.com/vega/vega-lite/issues/2415).';
+  'Axes cannot be shared in concatenated or concatenated views yet (https://github.com/vega/vega-lite/issues/2415).';
```
Comment (kanitw) [inline src/log/message.ts]: concatenated or concatenated ? lol

File: src/compile/unit.ts [hunk 11]
```diff
@@ -72,31 +71,21 @@ export class UnitModel extends ModelWithField {
...
 
-    const encodingWithRepeaterReplaced = replaceRepeaterInEncoding(spec.encoding ?? {}, repeater);
+    const encoding = spec.encoding ?? {};
 
-    this.markDef = initMarkdef(spec.mark, encodingWithRepeaterReplaced, config, {
+    this.markDef = initMarkdef(spec.mark, encoding, config, {
       graticule: spec.data && isGraticuleGenerator(spec.data)
     });
-    const encoding = (this.encoding = initEncoding(encodingWithRepeaterReplaced, this.markDef));
+    const normalizedEncoding = (this.encoding = initEncoding(encoding, this.markDef));
```
Comment (kanitw) [inline src/compile/unit.ts]: Let's not call this normalized as it doesn't happen in the normalized phase. Instead, just call this encoding and refer to the old `encoding` above via `spec.encoding`.

File: src/compile/selection/transforms/scales.ts [hunk 12]
```diff
@@ -34,12 +34,6 @@ const scaleBindings: TransformCompiler = {
       const extent = {selection: selCmpt.name, field: proj.field};
       scale.set('selectionExtent', extent, true);
       bound.push(proj);
-
-      // Bind both x/y for diag plot of repeated views.
-      if (model.repeater && model.repeater.row === model.repeater.column) {
```
Comment (kanitw) [inline src/compile/selection/transforms/scales.ts]: this seems problematic?

File: scripts/rename-schema.sh [hunk 13]
```diff
@@ -7,8 +7,8 @@ perl -pi -e s,'<Field\,','<',g build/vega-lite-schema.json
 perl -pi -e s,'<StandardType>','',g build/vega-lite-schema.json
 
 perl -pi -e s,'CompositeEncoding','Encoding',g build/vega-lite-schema.json
-perl -pi -e s,'Generic(.*)<FacetedUnitSpec\,LayerSpec>','\1',g build/vega-lite-schema.json
-
+perl -pi -e s,'Generic(.*)<FacetedUnitSpec\,LayerSpec\,?.*\,FieldName>','\1',g build/vega-lite-schema.json
+perl -pi -e s,'Generic(.*)<FacetedUnitSpec\,LayerSpec\,?.*\,Field>','Normalized\1',g build/vega-lite-schema.json
```
Comment (kanitw) [inline scripts/rename-schema.sh]: Can you explain why do we need this rename? Looks very complicated.

File: src/normalize/core.ts [hunk 14]
```diff
@@ -55,43 +58,80 @@ export class CoreNormalizer extends SpecMapper<NormalizerParams, FacetedUnitSpec
...
+          const childRepeater = {
+            repeat: repeatValue,
+            row: rowValue,
+            column: columnValue
+          };
+
+          const child = this.map(childSpec, {...params, repeater: childRepeater});
+
+          // TODO: remove child__repeat_ prefix
+          child.name =
```
Comment (kanitw) [inline src/normalize/core.ts]: This line is wrong, you should only include `child__repeat` only once in the beginning.  (Hence the name changes in the diff.)

File: examples/compiled/repeat_splom_cars.vg.json [hunk 15]
```diff
@@ -46,22 +45,25 @@
...
+  "signals": [
+    {"name": "childWidth", "value": 200},
+    {"name": "childHeight", "value": 200}
+  ],
   "layout": {"padding": 20, "columns": 2, "bounds": "full", "align": "all"},
   "marks": [
     {
       "type": "group",
-      "name": "child__repeat_row_Displacement__repeat_column_Horsepower_group",
+      "name": "child__repeat_row_Displacementchild__repeat_column_Horsepower_group",
```
Comment (kanitw) [inline examples/compiled/repeat_splom_cars.vg.json]: see https://github.com/vega/vega-lite/pull/6002#discussion_r390498678

File: examples/compiled/interactive_layered_crossfilter_discrete.vg.json (hunk 7 above)
Comment (domoritz) [reply examples/compiled/interactive_layered_crossfilter_discrete.vg.json]: Do you see an issue in the normalized (concat) spec? The height here is just the default, no?

File: test/compile/repeat.test.ts (hunk 8 above)
Comment (domoritz) [reply test/compile/repeat.test.ts]: We still need to replace repeater in encoding and facet in normalization. Should I move the test?

File: src/log/message.ts (hunk 10 above)
Comment (domoritz) [reply src/log/message.ts]: Ups. Will fix.

File: src/compile/selection/transforms/scales.ts (hunk 12 above)
Comment (domoritz) [reply src/compile/selection/transforms/scales.ts]: This is https://github.com/vega/vega-lite/issues/6000, right?

File: scripts/rename-schema.sh (hunk 13 above)
Comment (domoritz) [reply scripts/rename-schema.sh]: This is the similar to before. Just renaming generic specs with FieldName (string) to be normalized spec and generic spec with Field (string or repeat) to just be a spec.

File: test/compile/repeat.test.ts (hunk 8 above)
Comment (kanitw) [reply test/compile/repeat.test.ts]: Yes, to where the function is placed now.

File: src/compile/selection/transforms/scales.ts (hunk 12 above)
Comment (kanitw) [reply src/compile/selection/transforms/scales.ts]: Perhaps, but what's the effect of removing that?

File: test/compile/repeat.test.ts (hunk 8 above)
Comment (domoritz) [reply test/compile/repeat.test.ts]: I'm moving the files and tests then

File: src/compile/unit.ts (hunk 11 above)
Comment (domoritz) [reply src/compile/unit.ts]: Done

File: examples/compiled/interactive_panzoom_splom.vg.json [hunk 16]
```diff
@@ -1895,7 +1895,6 @@
       "name": "child__repeat_row_Horsepower__repeat_column_Horsepower_y",
       "type": "linear",
       "domain": {"data": "data_2", "field": "Horsepower"},
-      "domainRaw": {"signal": "grid[\"Horsepower\"]"},
```
Comment (kanitw) [inline examples/compiled/interactive_panzoom_splom.vg.json]: Is this change intentional? (and why?)

File: src/compile/selection/transforms/scales.ts (hunk 12 above)
Comment (domoritz) [reply src/compile/selection/transforms/scales.ts]: Let's try: https://github.com/vega/vega-lite/pull/6124

File: examples/compiled/interactive_panzoom_splom.vg.json (hunk 16 above)
Comment (domoritz) [reply examples/compiled/interactive_panzoom_splom.vg.json]: Answer is above.

File: src/normalize/core.ts [hunk 17]
```diff
@@ -55,43 +58,81 @@ export class CoreNormalizer extends SpecMapper<NormalizerParams, FacetedUnitSpec
...
+        }
+      }
+    }
+
+    const columns = isArray(repeat) ? spec.columns : repeat.column ? repeat.column.length : 1;
+
     return {
-      ...spec,
-      spec: this.map(spec.spec, params)
+      data: data ?? childSpec.data, // data from child spec should have precedence
```
Comment (kanitw) [inline src/normalize/core.ts]: The code here doesn't really give precedence to child spec?

File: src/normalize/core.ts [hunk 18]
```diff
@@ -55,43 +58,81 @@ export class CoreNormalizer extends SpecMapper<NormalizerParams, FacetedUnitSpec
...
+          };
+
+          const child = this.map(childSpec, {...params, repeater: childRepeater});
+
+          // TODO: remove child__repeat_ prefix
+          child.name =
+            'child' +
+            (repeatValue ? `__repeat_repeat_${varName(repeatValue)}` : '') +
+            (rowValue ? `__repeat_row_${varName(rowValue)}` : '') +
+            (columnValue ? `__repeat_column_${varName(columnValue)}` : '');
```
Comment (kanitw) [inline src/normalize/core.ts]: The naming logic here seems wrong for nested repeat? 

Existing repeatValue should be treated as prefix to the name, not inserted in the middle.
